Add Python dictionary -> Nim Table

Vindaar opened the pull request:
So this is still somewhat work in progress.

This PR adds support to convert Python dictionaries to:
- [x] homogeneous python dict -> Nim Table
- [ ] ~~heterogeneous** python dict -> Nim's `JsonNode`~~

Conversion of Python objects to `JsonNode`s will be done in a separate PR.

Vindaar commented:
Now added the ability to also convert Python lists popping up in dictionaries to `JsonNode.JArray`.

Two things are still missing:
- [ ] convert python tuple
- [ ] convert arbitrary Python object

The latter I'm not quite sure about. Depending on the object it becomes unwieldy.

And I just saw you already fixed the `nil` strings. I'll rebase this later.

Vindaar commented:
Ok, I rebased this onto the current master branch. Also implemented support for tuples -> `JsonNode` (are also converted to `JArray`). 
Python objects (for now), are simply converted to their string representation (which might not be very useful), but at least this way one can 
- convert a dictionary even if it contains objects
- at least see what object it contained.

I don't know if there's a smart way to deal with (at least the majority) of Python objects. So I'll consider this "done" for now. 

I still need to make the tests perform actual tests. Not sure if variable and proc names are fine yet (some of it seems to diverge from the rest I fear). And the `getVal` template, while useful does produce quite some code (I think about 400 lines of Nim code?). But I coulnd't come up with a more compact solution quickly.

yglukhov commented on nimpy.nim:
```diff
@@ -957,6 +1016,142 @@ proc pyObjToNimSeq[T](o: PPyObject, v: var seq[T]) =
+            actions
+        decRef pyVal
+
+proc `%`(x: Complex): JsonNode =
```
This shouldn't be needed

Vindaar replied:
If you mean the forward declaration, you're right. I simply didn't stage the deletion of it.

yglukhov commented on nimpy.nim:
```diff
@@ -957,6 +1016,142 @@ proc pyObjToNimSeq[T](o: PPyObject, v: var seq[T]) =
+proc `%`(o: PPyObject): JsonNode
+proc `%`(o: PyObject): JsonNode
+
+template getVal(pyVal: PPyObject, actions: untyped) {.dirty.} =
```
I suppose you dont need dirty here.

Vindaar replied:
Good point, I don't if I just inject `x`.

yglukhov commented on nimpy.nim:
```diff
@@ -899,6 +924,32 @@ proc newPyObject(o: PPyObject): PyObject =
     incRef o
     newPyObjectConsumingRef(o)
 
+proc checkPyObjPyType(o: PPyObject): PyBaseType =
```
Rename to `getBaseType` or `baseType`.

Vindaar replied:
Done.

yglukhov commented on nimpy.nim:
```diff
@@ -899,6 +924,32 @@ proc newPyObject(o: PPyObject): PyObject =
+                       pyLib.PyUnicode_Type : pbString,
+                       pyLib.PyList_Type : pbList,
+                       pyLib.PyTuple_Type : pbTuple,
+                       pyLib.PyDict_Type : pbDict }.toTable
```
No need in `toTable` here, and `checkTypes` can be const.

Vindaar replied:
Well, I guess I can iterate over the tuples and get the fields, true. It can't be const though, because `pyLib` isn't defined at compile time.

yglukhov commented on nimpy.nim:
```diff
@@ -934,12 +985,20 @@ proc pyObjToNim[T](o: PPyObject, v: var T) {.inline.} =
         pyObjToNimArray(o, v)
+    elif T is JsonNode:
+        # above `ref` since `JsonNode` is a `ref` object
+        conversionTypeCheck(pyLib.PyDict_Type)
```
Why only dict? It could be anything representable with JsonNode.

Vindaar replied:
Simply because I wanted to keep this PR "simple".

yglukhov commented on nimpy.nim:
```diff
@@ -957,6 +1016,142 @@ proc pyObjToNimSeq[T](o: PPyObject, v: var seq[T]) =
+        # for now we simply convert objects to their string
+        # representation. More useful than either crashing or
+        # dropping quietly
+        var x = % newPyObject(pyVal)
```
Can't you use `%` directly on `pyVal` here?

Vindaar replied:
With the current implementation, no (which is probably valid criticism), because I steal the `%` for `PPyObject` for Python dictionaries.

yglukhov commented on nimpy.nim:
```diff
@@ -957,6 +1016,142 @@ proc pyObjToNimSeq[T](o: PPyObject, v: var seq[T]) =
+    result = %* {"real" : x.re,
+                 "imag" : x.im}
+
+proc `%`(o: PyObject): JsonNode =
```
This shouldn't be needed

Vindaar replied:
I could certainly just do that in place. Seemed nicer to provide that proc for proper handling of objects in the future though.

yglukhov commented:
Sorry but your toJson conversion is a mess. Firstly i'd suggest separating it out to a separate pr. Secondly you'll have to rethink it:
- Every PPyObject is potentially convertible to JsonNode, not necessarily dict.
- There should be one recursive `toJson(PPyObject): JsonNode` proc that is just a switch-case by base type. No blowups expected.
Please feel free to ask should you need more details.

Vindaar commented:
No, I agree. The JSON conversion *did* turn out to be messy. It's just what I could come up with with my inexperience (especially regarding the Python C API, but also in general).

Yeah, I can strip out the JSON aspect here and try to rewrite it. Maybe not focusing on Python dictionaries will actually make thinking about JSON conversion easier. Thanks, I'll ask if I get stuck.

Vindaar commented:
@yglukhov Sorry it took me this long to get back to this. 

I removed everything but the conversion of dictionary -> Table now and cleaned up the test case. Currently if a datatype cannot be converted to the desired Nim types, we fail and raise in `pyObjToNim`. Do you think that's acceptable?

yglukhov commented:
> Do you think that's acceptable?

Yes, thank you!
